compiler_gym/service/client_service_compiler_env.py

```diff
@@ -0,0 +1,1291 @@
+from compiler_gym.service.proto import AddBenchmarkRequest
+from compiler_gym.service.proto import Benchmark as BenchmarkProto
+from compiler_gym.service.proto import (
+    EndSessionReply,
+    EndSessionRequest,
+    Event,
+    ForkSessionReply,
+    ForkSessionRequest,
+    GetVersionReply,
+    GetVersionRequest,
+    SendSessionParameterReply,
+    SendSessionParameterRequest,
+    SessionParameter,
+    StartSessionRequest,
+    StepReply,
+    StepRequest,
+    proto_to_action_space,
+)
+from compiler_gym.spaces import DefaultRewardFromObservation, NamedDiscrete, Reward
+from compiler_gym.util.gym_type_hints import (
+    ActionType,
+    ObservationType,
+    RewardType,
+    StepType,
+)
+from compiler_gym.util.shell_format import plural
+from compiler_gym.util.timer import Timer
+from compiler_gym.validation_error import ValidationError
+from compiler_gym.validation_result import ValidationResult
+from compiler_gym.views import ObservationSpaceSpec, ObservationView, RewardView
+
+logger = logging.getLogger(__name__)
+
+# NOTE(cummins): This is only required to prevent a name conflict with the now
+# deprecated ClientServiceCompilerEnv.logger attribute. This can be removed once the logger
+# attribute is removed, scheduled for release 0.2.3.
+_logger = logger
+
+
+def _wrapped_step(
+    service: CompilerGymServiceConnection, request: StepRequest
+) -> StepReply:
+    """Call the Step() RPC endpoint."""
+    try:
+        return service(service.stub.Step, request)
+    except FileNotFoundError as e:
+        if str(e).startswith("Session not found"):
+            raise SessionNotFound(str(e))
+        raise
+
+
+class ClientServiceCompilerEnv(CompilerEnv):
+    """Implementation using gRPC for a client-server communication.
+
+    :ivar service: A connection to the underlying compiler service.
+
+    :vartype service: compiler_gym.service.CompilerGymServiceConnection
+
+    :ivar action_spaces: A list of supported action space names.
+
+    :vartype action_spaces: List[str]
+
+    :ivar actions: The list of actions that have been performed since the
+        previous call to :func:`reset`.
+
+    :vartype actions: List[ActionType]
+
+    :ivar reward_range: A tuple indicating the range of reward values. Default
+        range is (-inf, +inf).
+
+    :vartype reward_range: Tuple[float, float]
+
+    :ivar observation: A view of the available observation spaces that permits
+        on-demand computation of observations.
+
+    :vartype observation: compiler_gym.views.ObservationView
+
+    :ivar reward: A view of the available reward spaces that permits on-demand
+        computation of rewards.
+
+    :vartype reward: compiler_gym.views.RewardView
+
+    :ivar episode_reward: If :func:`ClientServiceCompilerEnv.reward_space
+        <compiler_gym.envs.CompilerGym.reward_space>` is set, this value is the
+        sum of all rewards for the current episode.
+
+    :vartype episode_reward: float
```

Comment:
I think each of these `ivar` comments should a docstring on the corresponding abstract property in `CompilerEnv`.

Reply:
done